### ISSUE TITLE
Unable to call map on array of variable depth

### ISSUE THREAD

**Issue by danvk** [microsoft/TypeScript#35045]
I'm trying to write a function which transforms the coordinates of any type of GeoJSON geometry. The coordinates of a Geometry can be either `number[]` (a Point), `number[][]` (a LineString), `number[][][]` (a Polygon) or `number[][][][]` (a MultiPolygon). I have a recursive implementation which works but does not type check.

**TypeScript Version:**  3.7.2

**Search Terms:**
- 2349

**Code**

```ts
type Position = number[];

function isNestedArray<T extends any[]>(x: T | T[]): x is T[] {
  return Array.isArray(x[0]);
}

export function transformCoordArray(
  coords: Position | Position[] | Position[][],
  fn: (pt: Position) => Position
): any[] {
  if (isNestedArray(coords)) {
    coords;  // type is Position[] | Position[][]
    return coords.map(c => transformCoordArray(c, fn));
    //            ~~~ This expression is not callable.
  } else {
    coords;  // type is Position
    return fn(coords);
  }
}
```

([playground][bad])

The full error message is:

```
This expression is not callable.
  Each member of the union type '
   (<U>(callbackfn: (value: Position, index: number, array: Position[]) => U, thisArg?: any) => U[])
 | (<U>(callbackfn: (value: Position[], index: number, array: Position[][]) => U, thisArg?: any) => U[])
' has signatures, but none of those signatures are compatible with each other.(2349)
```

The call is valid because `c`'s type should be `Position | Position[]`, which is assignable to `Position | Position[] | Position[][]`.

Interestingly, I can make the error go away by assigning to a new variable which has a broader type (`(Position | Position[])[]`): 
([playground][ok])

```ts
export function transformCoordArray(
  coords: Position | Position[] | Position[][],
  fn: (pt: Position) => Position
): any[] {
  if (isNestedArray(coords)) {
    coords;  // type is Position[] | Position[][]
    const c2: (Position | Position[])[] = coords;
    return c2.map(c => transformCoordArray(c, fn));  // ok
  } else {
    coords;  // type is Position
    return fn(coords);
  }
}
```

**Expected behavior:**

No error.

**Actual behavior:**

The given error.

**Playground Link:** [playground][bad]

**Related Issues:** 

[bad]: https://www.typescriptlang.org/play/?ssl=14&ssc=55&pln=14&pc=1#code/C4TwDgpgBACg9gZwJbCXAdlAvFdBXAWwCMIAnAbQF0BuAWACgGAzPdAY1QyiQQDkIEwCABMAgqVIBDEAB4AKlAgAPIemEIok9CCoA+ABRKAXFAUAfU1QCUJpdw1yqUAN4MoUUhGB5SmcVJAAOh5-aUNyAAZKKzp6AF8GBmUwOFJgKBZ2TkxgKXQEJlSCAGE4VLEJMLcoNjLSdRN4ZGyoCyaUNHQnNsQOjCoqABpqpnQTfTBgRt7sq2xdWBnOhhtNbSdXenckJih9Hn5BEVCQfVryhCs5zfd3c-qEancAemeoUEh7RebO7u++rqUKjVdyeby+Gp1dSBAiSMBnebvPIFIqlconM6DDLoK6xW5QV74om3AB+ZNMAAseIolGBPAhkFxqeg4Ok2JIADYcyREDkQQLVOKKDkIaA3W73dRPAlvD7QantbIgjxeHyYUZnKGXPFQBLxBhAA
[ok]: https://www.typescriptlang.org/play/?ssl=14&ssc=1&pln=15&pc=1#code/C4TwDgpgBACg9gZwJbCXAdlAvFdBXAWwCMIAnAbQF0BuAWACgGAzPdAY1QyiQQDkIEwCABMAgqVIBDEAB4AKlAgAPIemEIok9CCoA+ABRKAXFAUAfU1QCUJpdw1yqUAN4MoUUhGB5SmcVJAAOh5-aUNyAAZKKzp6AF8GBmUwOFJgKBZ2TkxgKXQEJlSCAGE4VLEJMLcoNjLSdRN4ZGyoCyaUNHQnNsQOjCoqABpqpnQTfTBgRt7sq2xdWBnOhhtNbSdXenckJih9Hn5BEVCQfVryhCs5zfd3c-qEancAemeoUEh7RebO7u++rqUKjVO4YQQ1ABM43aLR6P360ScOHu6lit1unm8vkhgQIkjAZ3m7zyBSKpXKJzOgwy6CuTygrygcAA1tU4ooADYIaA3W4ox4vN4faA8f7ZEEeLw+TCjM51dQxNkMBL0IA

**Comment by jcalz** [comment]
Duplicate of #7294 (see [comment](https://github.com/Microsoft/TypeScript/issues/7294#issuecomment-465801362)).  Related to #29011.  

Mentioned in documentation as [caveats for improved behavior for calling union types](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-3.html#caveats):

> This new behavior only kicks in when at most one type in the union has multiple overloads, and at most one type in the union has a generic signature. That means methods on `number[] | string[]` like `map` (which is generic) still won’t be callable.

**Comment by thorn0** [comment]
While methods like `push` are obviously problematic for `number[] | string[]`, it looks safe to allow methods like `map` or `every` for `readonly number[] | readonly string[]`. Has this (special treatment of read-only arrays when merging union signatures) already been discussed in some issue?

**Comment by thorn0** [comment]
@RyanCavanaugh We are trying to gradually introduce type checking in the [Prettier](https://github.com/prettier/prettier) project and facing this issue. 

Prettier works with strongly typed ASTs that come from different parsers.

E.g. suppose we have an object of type `import('estree').Node | import('@typescript-eslint/typescript-estree').TSESTree.Node` and code like this:

```js
if (node.type === "ObjectExpression") {
    return node.properties.every(
      p => ...
    );
  }
```

Even though `node.properties` are correctly inferred to be of type `ObjectLiteralElementLike[] | Property[]` (because both `estree` and `@typescript-eslint/typescript-estree` define `ObjectExpression`, not exactly compatible though), `p` in the `every` callback is of type `any`. 😢

**Comment by hinell** [comment]
I think it's inherently bad idea to rely on compiler to figure out the output of the `transformCoordArray()` meanwhile `coords: ...` argument varies.
I think much better way either to create [overloads (checkout playground)][ts-example-0]:

```ts
  ...
  export function transformCoordArray(coords: Position[][], fn: PosTransfCb): Position[][]
  export function transformCoordArray(coords: Position[], fn: PosTransfCb): Position[]
  export function transformCoordArray(coords: Position, fn: PosTransfCb): Position
  ...
```

or assert `coords`:

```ts
   (<any> coords).map(c => transformCoordArray(c, fn)) as Position | Position[] | Position[][];
```

[ts-example-0]: https://www.typescriptlang.org/play/#code/HYQwtgpgzgDiDGEAEAZAlgIyQbwLACgkkIAPGAewCcAXJagTxmQAVyo1q1zgkBeJYAFcwGCJQDaAXQDcBAkQBmg4PE7ckaKADlo1CABMAgpUoh6AHgAqxEnuD6oSEMHpSAfAAoSALiTWAPn5SAJS+JBqOllI48kRIlBDUgpQ8xqb0AHSaaWZe4gAMksGyhEgAvrEMTEisUJamwFAKAMJY-B4w1L61HFzAwXxuNWy93CWxpBQ0SEoqajzUDU1UYM3kVEYmufDrlA7dI-NSUgA0M8AHdUstGKHD7EeSUiVEk1S0s6p9dNcraxs5egeHYbKCXUbAU7nS71ZxNVp3HqPCZkd4zZRfdSLOEKP67TbpYH4sH3CFnBQXe6wxo3RGHPooqYfDHzH44vEArZAkF7EnOejkym1anwrB3fkxUpENAKJAeTQ6KB6AnbYnBAZ4KVxHkOaREAD0+rojGQmlJjyQgSRfWOklicXiiWSPB1UAyYBAMGBgzZNI5e0BwMF6vtcUNDojcQAfjG-AALM2TBJQdjqM3Aci0eAgAA2OZAGBzEAyobKxBzUGQmsjrr1SHDVVNjmt3FDRASSRS5yJoIGIGb9LGpYIFXwcnwo4IADcQJQkOF+OgMBlsX7KKt8YHxABWbdnAAsAGZJGcOgNeEMYMECOFwy3gOYd3ukEfJG4gA

**Comment by jcalz** [comment]
Fixed by #42620 I think?

**Comment by danvk** [comment]
Yep! The error in the example in the original issue goes away with 4.3: [playground](https://www.typescriptlang.org/play?ts=4.3.0-beta&ssl=1&ssc=1&pln=2&pc=1&strict=true#code/C4TwDgpgBACg9gZwJbCXAdlAvFdBXAWwCMIAnAbQF0BuAWACgGAzPdAY1QyiQQDkIEwCABMAgqVIBDEAB4AKlAgAPIemEIok9CCoA+ABRKAXFAUAfU1QCUJpdw1yqUAN4MoUUhGB5SmcVJAAOh5-aUNyAAZKKzp6AF8GBmUwOFJgKBZ2TkxgKXQEJlSCAGE4VLEJMLcoNjLSdRN4ZGyoCyaUNHQnNsQOjCoqABpqpnQTfTBgRt7sq2xdWBnOhhtNbSdXenckJih9Hn5BEVCQfVryhCs5zfd3c-qEancAemeoUEh7RebO7u++rqUKjVdyeby+Gp1dSBAiSMBnebvPIFIqlconM6DDLoK6xW5QV74om3AB+ZNMAAseIolGBPAhkFxqeg4Ok2JIADYcyREDkQQLVOKKDkIaA3W73dRPAlvD7QantbIgjxeHyYUZnKGXPFQBLxBhAA)
